File: floating_palette/macos/Classes/services/MessageService.swift
import Cocoa
import FlutterMacOS

/// Handles host ↔ palette messaging.
final class MessageService {
    private let store = WindowStore.shared
    private var eventSink: ((String, String, String?, [String: Any]) -> Void)?

    func setEventSink(_ sink: @escaping (String, String, String?, [String: Any]) -> Void) {
        self.eventSink = sink
    }

    // MARK: - Commands

    func handle(_ command: String, windowId: String?, params: [String: Any], result: @escaping FlutterResult) {
        switch command {
        case "send":
            send(windowId: windowId, params: params, result: result)
        default:
            result(FlutterError(code: "UNKNOWN_COMMAND", message: "Unknown message command: \(command)", details: nil))
        }
    }

    // MARK: - Send to Palette (Host → Palette)

    private func send(windowId: String?, params: [String: Any], result: @escaping FlutterResult) {
        guard let id = windowId, let window = store.get(id) else {
            result(FlutterError(code: "NOT_FOUND", message: "Window not found", details: nil))
            return
        }

        guard let channel = window.messengerChannel else {
            result(FlutterError(code: "NO_CHANNEL", message: "Messenger channel not available", details: nil))
            return
        }

        let type = params["type"] as? String ?? ""
        let data = params["data"] as? [String: Any] ?? [:]

        DispatchQueue.main.async {
            // Send to palette via its messenger channel
            channel.invokeMethod("receive", arguments: [
                "type": type,
                "data": data
            ])
            result(nil)
        }
    }
}

File: floating_palette/windows/services/message_service.cpp
#include "message_service.h"

#include "../core/logger.h"
#include "../core/param_helpers.h"

namespace floating_palette {

void MessageService::Handle(
    const std::string& command,
    const std::string* window_id,
    const flutter::EncodableMap& params,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  if (command == "send") {
    Send(window_id, params, std::move(result));
  } else {
    result->Error("UNKNOWN_COMMAND", "Unknown message command: " + command);
  }
}

void MessageService::Send(
    const std::string* window_id,
    const flutter::EncodableMap& params,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  if (!window_id) {
    result->Error("MISSING_ID", "windowId required");
    return;
  }

  auto* window = WindowStore::Instance().Get(*window_id);
  if (!window) {
    result->Error("NOT_FOUND", "Window not found: " + *window_id);
    return;
  }

  // Forward message to the palette's messenger channel
  if (window->messenger_channel) {
    window->messenger_channel->InvokeMethod(
        "receive",
        std::make_unique<flutter::EncodableValue>(
            flutter::EncodableValue(params)));
  }

  result->Success(flutter::EncodableValue());
}

}  // namespace floating_palette
